ENH: add corrmap.

dengemann opened the pull request:
Taking over #1795

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -2093,3 +2110,228 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+            sim_i_o = np.abs(np.corrcoef(target, newtarget)[1, 0])
+
+            return newtarget, median_corr_with_target, sim_i_o, max_corrs
+        except IndexError:
```
Is there a more elegant way to do this? a problem with `try/except` over a bunch of lines is that sometimes you end up hitting an unintended error. It's also not clear from the code where or why an `IndexError` would occur, so it's also harder for the next person to deal with.

dengemann replied:
I know ... tried to brainwash @jona-sassenhagen but he was very resistant so far.

Eric89GXL replied:
I understand the "it's easier to ask for forgiveness then for permission" mentality, but it's not clear in this code what the forgiveness would even be for, which makes it difficult to understand. I think it's typically applied to a single or a couple lines of code. Looking at these ~10 lines of code deeper I still can't figure out where to expect an indexing error, which is not a good sign.

dengemann replied:
I agree. let's figure it out :)

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -2093,3 +2110,228 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+            components = list(range(ica.n_components_))
+        maps = fast_dot(ica.mixing_matrix_[:, components].T,
+                        ica.pca_components_[:ica.n_components_])
+        return maps
```
Why nest these functions instead of just making private functions?

dengemann replied:
because they aren't used by any other function in this module.

Eric89GXL replied:
It is inconsistent with how we have done everything else, we typically avoid nesting functions unless there is a good reason to do so...

dengemann replied:
we have a few instances of nested functions here and there. case law would not that clearly suggest not to do it. But I'm fine to change it if you insist (allthough I'm of course lazy).

Eric89GXL replied:
One of the reasons I don't like it is that it reduces readability and hinders code reuse.

If I want the answer to the question, "what functions does this file have that I could potentially reuse?", nesting functions makes that question (potentially much) more difficult to answer. Indentation makes it harder to see functions in the first place. And more importantly, nested functions can make use of variables within the enclosing function's scope. So now to figure out if I can reuse one of those nested functions, I have to look through *all* of the variables it uses to figure out if any of them belong to the parent function. Actually I generally assume that if a function is nested, it *does* make use of some local variable, because if it didn't... why would it need to be nested in the first place? You also get hit with a 4-character indentation/line-wrapping penalty, but that's a minor point :)

jona-sassenhagen replied:
I guess in the long run there should be a public function to get maps anyways.

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -2093,3 +2110,228 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+    else:
+        logger.info("At least 1 IC detected for each subject.")
+
+    return None if plot is False else (template_fig, labelled_ics)
```
One problem with using `if plot is True` above and `if plot is False` here, is that you cannot assume they are complementary, e.g. `2 is True` is `False` and `2 is False` is `False`. It should be safer to just use `if plot:` above and `if not plot` here, because those are complementary.

Eric89GXL commented on mne/tests/test_utils.py:
```diff
@@ -414,6 +415,28 @@ def test_check_type_picks():
 
 
+def test_compute_corr():
+    """Anscombe's Quartett
```
Test Anscombe's...

Eric89GXL commented on mne/tests/test_utils.py:
```diff
@@ -414,6 +415,28 @@ def test_check_type_picks():
+                   for i in range(len(y))]).round(3)
+    assert(r[3] != r[0] == r[1] == r[2] != r[4])
+    for r_v, r2_v in zip(r, r2):
+        assert(r_v == r2_v)
```
assert_equal here, assert_true above

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -45,6 +50,8 @@
 from .ctps_ import ctps
 from ..externals.six import string_types, text_type
 
+from ..utils import compute_corr
```
you should import `fast_dot` now, as well, and remove it from here (DRY)

Eric89GXL commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -268,7 +274,16 @@ def test_ica_additional():
+        ica.fit(raw, picks=[1, 2, 3, 4, 5], start=start, stop=stop2)
+
+    # test corrmap
+    ica2 = deepcopy(ica)
```
do we have `ica.copy()`? If not, we probably should. If so, it should be used here instead.

Eric89GXL commented:
Other than my comments, LGTM, but I don't really know much about the ICA code

dengemann commented:
@Eric89GXL bad news the gamma map is not happy without additional regularization

https://www.dropbox.com/s/mtlx9ube6m8iza5/Screenshot%202015-04-20%2004.35.22.png?dl=0

dengemann commented:
deleted comment, wrong issue.

agramfort commented:
do you want this merged for the release?

Eric89GXL commented:
@agramfort #1795, the PR that this one superceded, was marked for 0.9 so I think yeah he wants it for the release

dengemann commented:
> do you want this merged for the release?

let's try to do that.

Eric89GXL commented:
@dengemann don't forget about this one

dengemann commented:
I'm hesitating / thinking about moving this to 0.10. Let me see how this weeks schedule develops (unless @jona-sassenhagen want's to complete it). At this stage it's more important though to fix bugs than adding new features.

jona-sassenhagen commented:
Is there anything major missing? I wouldn't stress it being a .9 feature, but I'd like to have it merged so the people I work with have an easier time using it :)

dengemann commented:
> Is there anything major missing? I wouldn't stress it being a .9 feature, but I'd like to have it merged so the people I work with have an easier time using it :)

Nope just addressing the comments.

- [ ] move nested functions to private functions at the module level
- [ ] remove try catch blocks

dengemann commented:
and some minor points.

jona-sassenhagen commented:
I have a bit of time now, I'll try to get at least one of the big two

jona-sassenhagen commented:
that is, this week, not right now :)

Eric89GXL commented:
I'll just do it now, they're my gripes anyway :)

jona-sassenhagen commented:
better still :)

dengemann commented:
@Eric89GXL you do have push access. Feel free to work on my PR.

Eric89GXL commented:
done, @dengemann and/or @jona-sassenhagen please have a look. I got rid of the try/excepts other than the one I thought made sense (the `read_meas_info` one is a good example of the ask for forgiveness principle IMO) by checking for failure conditions. Note that this is also a better approach because it avoided a spewing of `numpy` warnings in some cases before hitting the eventual `ValueError`.

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -933,7 +935,10 @@ def find_bads_ecg(self, inst, ch_name=None, threshold=None,
         ecg_idx = ecg_idx[np.abs(scores[ecg_idx]).argsort()[::-1]]
-        return list(ecg_idx), scores
+        if not hasattr(self, 'labels_'):
+            self.labels_ = {}
```
{} -> dict()

it's more explicit

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -1018,7 +1023,10 @@ def find_bads_eog(self, inst, ch_name=None, threshold=3.0,
 
-        return eog_idx, scores
+        if not hasattr(self, 'labels_'):
+            self.labels_ = {}
```
same here

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -2102,3 +2121,235 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+    template_fig : fig
+        Figure showing the mean template.
+    labelled_ics : fig
+        Figure showing the labelled ICs per subject.
```
subject?

per subject -> in all ica decompositions.

jona-sassenhagen replied:
Agree. In the example, we even use Corrmap do label across multiple decompositions of the same subject.

agramfort commented on mne/tests/test_utils.py:
```diff
@@ -414,6 +415,29 @@ def test_check_type_picks():
+
+    r = compute_corr(x, y.T).round(3)
+    r2 = np.array([np.corrcoef(x, y[i])[0, 1]
+                   for i in range(len(y))]).round(3)
```
why the round(3) ?

jona-sassenhagen commented on mne/preprocessing/ica.py:
```diff
@@ -1018,7 +1023,10 @@ def find_bads_eog(self, inst, ch_name=None, threshold=3.0,
+        if not hasattr(self, 'labels_'):
+            self.labels_ = {}
+        self.labels_['eog'] = list(eog_idx)
+        return self.labels_['eog'], scores
```
corrmap defaults to 'bads'. I'm not sure about this, but maybe it'd make sense to unify all stuff that's primarily useful for a latter call to ica.apply( ... exclude= ...) under one default label.

Eric89GXL replied:
I'm not sure what you mean @jona-sassenhagen, I assume this is a comment for @dengemann. Or at least I'll let you two sort it out :)

dengemann replied:
So this is the method for finding eog artifacts, so we should have an expplicit label for that. if it corresponds to the family of bads as defined by another method that's fine, no?

jona-sassenhagen commented:
Looks good from here, I'll give it a try on real data tomorrow.

Eric89GXL commented:
Addressed @agramfort's comments. Okay from your end @agramfort?

dengemann commented:
yo, cleanup + support for custom outlines.

dengemann commented:
let's see if travis is happy

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -2102,3 +2121,275 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+                                     contours=contours, layout=layout,
+                                     show=show)
+
+    return None if plot is False else template_fig, labelled_ics
```
You have missed the point, I'm afraid. What will happen here if a user passes `plot=None` or `0.` or `[]` or `''` instead of either `True` or `False`? They will get some cryptic error about how the variables `template_fig` and `labelled_ics` do not exist. I am advocating *not* to use `is` in either case, because when you use it, you are making an implicit assumption that `x is True` and `x is False` is complementary, and they are not.

dengemann replied:
... well we rely on this pattern almost everywhere ... the only safe way to deal with it would be to check bool types everywhere ....

Eric89GXL replied:
... or `plot=1.` or `plot=1` or `'yes'` or `[1]` will also have the same error. Basically if they pass anything other than `True` or `False`, they will get a cryptic error here, whereas checking `if x` and `if not x` will not suffer from this problem.

Eric89GXL replied:
I recognize that `x is True` and `x is False` are useful in some cases, e.g. where you want to be able to differentiate between `x is None` and `len(x) == 0` and `x == 0` and `not x`, but this doesn't seem to be one of those cases...

dengemann replied:
The thing is that we follow the `is` convention for quite a long time.
Look at git grep "is True" and "is False"

dengemann replied:
the thins is that via doc the API contract says bool, not list, not str, not int, etc.

Eric89GXL replied:
In some cases we are using it correctly, e.g. when a parameter can be `True | False | None` you need to do it. But if it's just a boolean check, you probably shouldn't. In that sense, I'm arguing that our use elsewhere is probably also incorrect.

dengemann replied:
... oh boy. Maybe separate issue + discussion?

dengemann replied:
if you contends that it's incorrect it's incorrect in *many* places.

Eric89GXL replied:
> the only safe way to deal with it would be to check bool types everywhere

I am arguing that a safe way to deal with this problem is not to check bools everywhere, but to use `if x` and `if not x` everywhere that it's appropriate instead.

dengemann replied:
.... but the doc says bool. You're just not supposed to pass anything else than True or False.

Eric89GXL replied:
Yeah, I think it might be incorrect in many places. I'm just trying to save users some pain. I could see people being lazy and using 0 or 1, and getting a weird error, and having no idea why. And I'm convinced they even should get an error, when what they passed would clearly evaluate to `True` or `False` if you checked `bool(x)`...

Eric89GXL replied:
I agree the doc says bool. But first, we cannot control what users pass, only try to control the damage and confusion they cause :) Second, I don't think it's that unreasonable for a user to assume under most circumstances that by "bool" someone generally means something that evaluates as True or False appropriately with `bool(x)`. Thinking about `numpy` and `scipy` packages, I highly doubt in most cases if they have a parameter listed as `bool` they will throw an error if you use `0` or `1` instead.

dengemann replied:
Sounds like the perl way, allowing for any type. I personally dislike a bit the fact that anything can be bool. With bool checking I meant to have a function that only validates True and False and throws an error for anything else.

Eric89GXL replied:
Since we've followed this pattern you don't have to change anything here, I just don't want to hear any complaints about having to change it if a user getting some strange error down the line :)

dengemann replied:
Yeah I know, I'm just too lazy now. I have thought about this often, maybe it's a good momen to discuss this and agree explicitly on one approach. Maybe @agramfort also want's to contribute his 2 cents. Once we're on one page we should implement it consistently everywhere.

jona-sassenhagen replied:
How about, for a pragmatic solution just in this specific instance, add a line to the beginning ```if not isinstance(plot, bool): blah...```?

Or here:

```
if plot is True:
    return template_fig, labelled_ics
elif plot is False:
    return None
else:
    blah
```

agramfort replied:
I would do

if not isinstance(plot, bool): raise ValueError ...

dengemann commented:
@jona-sassenhagen apart from that general debate, are we good?

agramfort commented on examples/preprocessing/plot_corrmap_detection.py:
```diff
@@ -0,0 +1,74 @@
+# here we create artificial subsets
+
+all_epochs = mne.Epochs(raw, events, event_id, tmin, tmax,
+                        proj=False, picks=picks, baseline=(None, 0),
```
why do you baseline if you highpassed above?

dengemann replied:
good point

agramfort commented on examples/preprocessing/plot_corrmap_detection.py:
```diff
@@ -0,0 +1,74 @@
+#    to a pre-specified template across all subsets
+#    (or, in the real world, multiple participant data sets)
+
+template = (0, 0)
```
how do you know that the first component will be of type blink?

dengemann replied:
by inspection. this is semi-automatic.

agramfort replied:
>
> by inspection. this is semi-automatic.
>

can you add a comment then.

jona-sassenhagen replied:
Maybe as simple as

```
# 2) Use corrmap to identify the maps best corresponding
#    to a manually selected pre-specified template across all 
#    subsets (or, in the real world, multiple participant data sets)
```

agramfort replied:
ok but then you write label='blinks' and this seems coming out of nowhere? if people copy paste this example they will expect it to find the blinks which is not guaranteed...

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -2102,3 +2121,275 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+        A list of fitted ICA objects.
+    template : tuple
+        A tuple with two elements (int, int) representing the list indices of
+        the set from which the template should be chosen, and the template.
```
can you add an example.

If template is (1, 0) then we use the first component of the second ica instance.

is that correct?

jona-sassenhagen replied:
Yes.

agramfort replied:
can you this example to docstring then?

jona-sassenhagen commented:
@dengemann I don't see anything bad.

agramfort commented:
you'll need to rebase

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -2102,3 +2121,275 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+        used (via .add_artist). Defaults to True.
+    contours : int | False | None
+        The number of contour lines to draw. If 0, no contours will be drawn.
+
```
rm empty line

agramfort commented on mne/tests/test_utils.py:
```diff
@@ -414,6 +415,27 @@ def test_check_type_picks():
+    r = compute_corr(x, y.T)
+    r2 = np.array([np.corrcoef(x, y[i])[0, 1]
+                   for i in range(len(y))])
+    assert_allclose(r, r2, atol=1e-3)
```
why only an atol of 1e-3? I would expect a much better numerical accuracy for this

wmvanvliet commented:
the current implementation breaks if there are more then 20 IC's being plotted:

    /Users/rodin/BRU/mne-python/mne/preprocessing/ica.py in _plot_corrmap(data, subjs, indices, ch_type, ica, label, show, outlines, layout, cmap, contours)
       2189         figs = [_plot_corrmap(data[k:k + p], subjs[k:k + p],
       2190                 indices[k:k + p], ch_type, ica, label)
    -> 2191                 for k in range(0, n_components, p)]
       2192         return figs
       2193     elif np.isscalar(picks):
      
    TypeError: _plot_corrmap() takes exactly 11 arguments (6 given)

jona-sassenhagen commented on mne/preprocessing/ica.py:
```diff
@@ -2102,3 +2121,275 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+    """Customized ica.plot_components for corrmap"""
+    import matplotlib.pyplot as plt
+
+    title = 'Detected components of type ' + label
```
Label defaults to None and should stay None for dry runs, in which case we get an error here.

Eric89GXL commented:
@dengemann should we move this to 0.10, or do you have time to finish it?

jona-sassenhagen commented:
Would it help if I made proposals/PRs for the problems noted by @wmvanvliet and I?

dengemann commented:
As said before please take over if you have time.

> On 29 Apr 2015, at 13:23, jona-sassenhagen <notifications@github.com> wrote:
> 
> Would it help if I made proposals/PRs for the problems noted by @wmvanvliet and I?
> 
> —
> Reply to this email directly or view it on GitHub.
>

agramfort commented:
I would move this to 0.10... but you can object...

jona-sassenhagen commented:
@agramfort I think this is basically done.

jona-sassenhagen commented:
I have a PR addressing the problems noted by wmvanvliet and I.

Eric89GXL commented:
After that is merged it will need a rebase, too

Eric89GXL commented:
@dengemann if you are swamped I can merge the PR into your branch and rebase

agramfort commented:
my 0.5 concern is that I don't know if this method is what we should
advertise and it's currently the best we can offer. The problem when adding
code to a package is that it never goes away so we need to think twice...

dengemann commented:
Agreed. Let's move slowly, test it extensivel, and keep it on master for
now.

2015-05-12 21:59 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

> my 0.5 concern is that I don't know if this method is what we should
> advertise and it's currently the best we can offer. The problem when adding
> code to a package is that it never goes away so we need to think twice...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1985#issuecomment-101402889>
> .
>

jona-sassenhagen commented:
On master instead of release? Fine with me.

@dengemann, I'm up for a massive test comparing different ICA and IC identification mechanisms on many data sets ...

agramfort commented:
I would vote not to merge this to master before someone does an big
validation like you proposed. If we merge to master it will be released in
the following release

dengemann commented:
I meant merge it to 0.10 master once it's open and once we're happy.

2015-05-12 22:13 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

> I would vote not to merge this to master before someone does an big
> validation like you proposed. If we merge to master it will be released in
> the following release
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1985#issuecomment-101407018>
> .
>

jona-sassenhagen commented:
I've finally fixed the stuff that went broke in the meantime with updates on master (e.g. giving everything a show keyword destroyed the plotting), should I make a new PR or PR to @dengemann 's repo?..

agramfort commented:
Make a new PR to take the lead and get some free time to @dengemann ;)

jona-sassenhagen commented:
Close for #2104?
